### PR TITLE
Added JBoss Forge from scratch quickstart

### PR DESCRIPTION
**Pull request by lincolnthree**

### CODE REVIEW

**Comment by sgilda** [comment]
Hi Lincoln, 
I'm sorry, but I'm fairly new to git and am having a problem finding your quickstart. 
Is this the correct repository?  https://github.com/lincolnthree/forge.git
I can see from your commits that there is a forge-from-scratch directory, but I don't see it in your github or in the branch I created from it.
Can you tell me what I'm doing wrong?
Thanks,
Sande

**Comment by lincolnthree** [comment]
Nope, that's not it. You can see the repository the file came from in the "Commits" or "Diff" section of this request.

https://github.com/forge/quickstarts

Cheers,
~Lincoln

**Comment by sgilda** [comment]
Never mind! Wrong repo. Found it. Thanks!

**Comment by sgilda** [comment]
Hi Lincoln,
I'm following the directions in your quickstart. I downloaded and set up forge using the link you provided. I opened a shell prompt and ran 'forge' and it started fine. 
Your instructions at this point say:     Notice that there is a file called `generate.fsh` in this directory; run from Forge
    using the `run` command:
         $ run generate.fsh

Since I'm ran forge from my home directory, there is no generate.fsh file at that location. Should the instructions note to change to the quickstart/forge-from-scratch directory? 

Being picky, but there's another prompt you might want to add: Use [/home/sgilda/Quickstarts/quickstart/forge-from-scratch/example] as project directory? [Y/n]

When I got to the instructions for build and deploy, I ran into this problem:
    [example] example $ as7 deploy
    **_ERROR**_ [as7 deploy] $JBOSS_HOME must be set before deployment can continue.

My understanding is that we are not supposed to set up a JBOSS_HOME environment variable with AS7/EAP6. Or am I wrong here?

**Comment by lincolnthree** [comment]
Ah. Yes. We should note that 'cd quickstart-dir/...' might be necessary, or
we can tell them to supply the fully qualified path to the file from
wherever they would like to create the project.

Changing that prompt is not really practical at this point.

Regarding AS7 JBOSS_HOME, it's Forge that requires this, not AS7 itself. We
can work on that or suggest another deployment option like copying the file?

~Lincoln

On Thu, Jan 12, 2012 at 5:24 PM, sgilda <
reply@reply.github.com

> wrote:
> 
> Hi Lincoln,
> I'm following the directions in your quickstart. I downloaded and set up
> forge using the link you provided. I opened a shell prompt and ran 'forge'
> and it started fine.
> Your instructions at this point say:     Notice that there is a file
> called `generate.fsh` in this directory; run from Forge
>    using the `run` command:
>         $ run generate.fsh
> 
> Since I'm ran forge from my home directory, there is no generate.fsh file
> at that location. Should the instructions note to change to the
> quickstart/forge-from-scratch directory?
> 
> Being picky, but there's another prompt you might want to add: Use
> [/home/sgilda/Quickstarts/quickstart/forge-from-scratch/example] as project
> directory? [Y/n]
> 
> When I got to the instructions for build and deploy, I ran into this
> problem:
>    [example] example $ as7 deploy
>    **_ERROR**_ [as7 deploy] $JBOSS_HOME must be set before deployment can
> continue.
> 
> My understanding is that we are not supposed to set up a JBOSS_HOME
> environment variable with AS7/EAP6. Or am I wrong here?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/jbossas/quickstart/pull/53#issuecomment-3471679

## 

Lincoln Baxter, III
http://ocpsoft.com
http://scrumshark.com
"Keep it Simple"

**Comment by sgilda** [comment]
I defined a JBOSS_HOME variable in the forge shell prompt just to get further along. I was able to deploy and  access the application. Very nice!
It might be nice to talk about the files that are generated by forge. Point them to the src/ directory, describe what's there. and maybe talk about how they relate to the commands in the generate.fsh file? Isn't that the power of forge?

**Comment by sgilda** [comment]
Maybe we should tell them to define the variable in the prompt where 
they'll be running forge?

**Comment by pmuir** [comment]
re JBOSS_HOME, all we need to do is make sure that we tell them how to set this in the README, it's not a general requirement for the quickstarts.

Or, the generated project could include the jboss-as maven deploy plugin which doesn't require this (in fact, I think we should generally switch over to that approach in forge, i'll file an issue).

File: forge-from-scratch/generate.fsh [hunk 1]
```diff
@@ -0,0 +1,49 @@
```
Comment (pmuir) [inline forge-from-scratch/generate.fsh:0]: I think we could make this file a lot more useful to users by adding a comment for each line to explain what forge is doing. A real beauty of scripting forge is that users can learn from the script, as there is no magic...

File: forge-from-scratch/readme.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
+Forge From Scratch
+========================
+
+What is it?
+-----------
+
+This is an example of creating a fully `Java EE compliant` project `using nothing
```
Comment (pmuir) [inline forge-from-scratch/readme.md]: back ticks in markdown give you a inline code element. I think you want emphasis or something here?

File: forge-from-scratch/generate.fsh (hunk 1 above)
Comment (sgilda) [inline forge-from-scratch/generate.fsh:0]: I agree with Pete. It's uneventful without it.

File: forge-from-scratch/readme.md (hunk 2 above)
Comment (lincolnthree) [inline forge-from-scratch/readme.md]: I used it intentionally, suggestions welcome. I'm new to markdown

File: forge-from-scratch/generate.fsh (hunk 1 above)
Comment (lincolnthree) [inline forge-from-scratch/generate.fsh:0]: Ok. Let me figure out how to do comments again :p

File: forge-from-scratch/readme.md (hunk 2 above)
Comment (pmuir) [inline forge-from-scratch/readme.md]: See http://daringfireball.net/projects/markdown/syntax - you probably want stars or underscores.

**Comment by lincolnthree** [comment]
Ok. This is ready for another look.

File: forge-from-scratch/readme.md (hunk 2 above)
Comment (lincolnthree) [inline forge-from-scratch/readme.md]: Thanks!

File: forge-from-scratch/readme.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  
+or if you are using windows
+ 
+    $JBOSS_HOME/bin/standalone.bat
+
+To build the application, type 'build', then to deploy the application, use the 
+'jboss-as-7' Forge plugin; just type:
+
+     $ forge install-plugin jboss-as-7
+	 $ as7 setup
```
Comment (pmuir) [inline forge-from-scratch/readme.md]: I'm having problems with this bit. I get https://gist.github.com/1626475 when installing and a problem with setup

```
[acme] acme $ as7 setup
***ERROR*** [as7] Missing command for plugin [as7], available commands: [undeploy, deploy, deployexploded]
```

**Comment by lincolnthree** [comment]
Ah yes. That was a bug that I fixed in the latest snapshots. I've created a tag in the jboss-as-7 plugin repository for Beta5 (which is the workaround.) It should work for you now (and in the future.) Just reinstall from the plugin repo.

~Lincoln

**Comment by pmuir** [comment]
Looking good. Once last problem -

```
[INFO] Scanning for projects...
[WARNING] 
[WARNING] Some problems were encountered while building the effective model for com.acme:acme:war:1.0.0- SNAPSHOT
[WARNING] 'build.plugins.plugin.version' for org.apache.maven.plugins:maven-war-plugin is missing. @ line 84, column 15
[WARNING] 
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING] 
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING] 
[INFO]      
```

**Comment by lincolnthree** [comment]
Good find. Fixed in the next version :)

https://issues.jboss.org/browse/FORGE-437

~Lincoln

On Wed, Jan 18, 2012 at 7:03 AM, Pete Muir <
reply@reply.github.com

> wrote:
> 
> Looking good. Once last problem -
> 
>    [INFO] Scanning for projects...
>    [WARNING]
>    [WARNING] Some problems were encountered while building the effective
> model for com.acme:acme:war:1.0.0- SNAPSHOT
>    [WARNING] 'build.plugins.plugin.version' for
> org.apache.maven.plugins:maven-war-plugin is missing. @ line 84, column 15
>    [WARNING]
>    [WARNING] It is highly recommended to fix these problems because they
> threaten the stability of your build.
>    [WARNING]
>    [WARNING] For this reason, future Maven versions might no longer
> support building such malformed projects.
>    [WARNING]
>    [INFO]
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/jbossas/quickstart/pull/53#issuecomment-3544646

## 

Lincoln Baxter, III
http://ocpsoft.com
http://scrumshark.com
"Keep it Simple"

**Comment by sgilda** [comment]
Hi Lincoln, 
I haven't had a change to test this again, but I do have a couple of suggestions, if you don't mind:
- Can you uppercase the name of the README.md file for consistency with the other quickstarts?
- Can you make the following URLs in the README file actual links, so people can click on them?
  https://docs.jboss.org/author/display/FORGE/Installation
  http://github.com/forge/plugin-openshift-express/
  I will try to test this first thing tomorrow. 
  Thanks, Sande

**Comment by sgilda** [comment]
A few more comments:
In the README, you say the following:
    Then, you need to run `$ forge` from the command line, or from within JBoss Tools
    or by pressing CTRL-4.
    Notice that there is a file called `generate.fsh` in this directory; run from Forge
    using the `run` command:
     $ run generate.fsh
If you aren't in the quickstart/forge-from-scratch directory, the generate.fsh isn't there. Don't you need to navigate to that directory?

Aslo, I don't see where you talk about the files that are generated when you run this command. 

It could be me. Maybe I'm looking at the wrong version of this file?

**Comment by mojavelinux** [comment]
I think you should consider using a separate prompt syntax in the tutorial to help readers understand when they are typing something from the Forge shell and when they are using their normal shell.

In a tutorial I wrote recently, I used the following prefix:

forge>

I know that Forge doesn't use this prompt synax, but I think most people will get that this means they should be in the Forge shell.

File: forge-from-scratch/generate.fsh [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+@/* Create a new project in the current directory */;
+new-project;
+
+@/* This means less typing. If a script is automated, or is not meant to be interactive, use this command */;
+set ACCEPT_DEFAULTS true;
+
+@/* Turn our Java project into a Web project with JSF, CDI, EJB, and JPA  */;
+scaffold setup --scaffoldType faces;
+persistence setup --provider HIBERNATE --container JBOSS_AS7 ;
```
Comment (mojavelinux) [inline forge-from-scratch/generate.fsh:14]: I think you should add Validation setup here.

The Validation plugin should also ask whether you want the Hibernate extensions just like the JPA plugin. Java EE purists are going to dislike that they ended up with the Hibernate Validation libraries if they didn't intend to add them.

Comment (lincolnthree) [inline forge-from-scratch/generate.fsh:14]: Do you need to do anything special to activate Bean Validation in EE6+ ? Config files, etc?

Comment (mojavelinux) [inline forge-from-scratch/generate.fsh:14]: Nope. Just works.

## 

Sent from my CyanogenMod-powered
Android device, an open platform for
carriers, developers and consumers.
On Jan 24, 2012 3:54 PM, "Lincoln Baxter, III" <
reply@reply.github.com>
wrote:

Comment (lincolnthree) [inline forge-from-scratch/generate.fsh:14]: We can focus on this once we get the initial version accepted into the quickstarts.

**Comment by lincolnthree** [comment]
changed to using `forge>` to represent forge commands. good call

**Comment by lincolnthree** [comment]
Ok. I added a new section which links to the UI Scaffolding guide. This is a lot of information, so I don't think it should necessarily be directly included in the Readme, but if you disagree, I can move some of it over.

**Comment by sgilda** [comment]
I really like the link to the UI Scaffolding guide. I think that is a great addition!

The only thing I'd like to see added is a note after you run generate.fsh to look in the forge-from-scratch folder to see the resulting application. That will make it clear that this is where the magic happens.
I would almost move the "What did we create?" there too.

For example:

Notice that there is a file called generate.fsh in this directory; run from Forge using the run command:
 forge> run generate.fsh
This command will prompt you for the {project-name} (E.g: 'example'), and will also prompt for the top level package. This should be the domain for your organization. (E.g: 'com.example')

What did we create?
This quickstart has set up a native Java EE 6 application. After this command completes, you will see a folder with the same name as the project-name you entered in your <quickstart>/forge-from-scratch folder. Browse through this project to see the code that was generated as a result of this command. For a full description of what was generated with this script, and the structure of the application, you should visit the Forge UI Scaffolding Guide

Deploy and test the newly generated application
Next, you need to start JBoss AS 7 (or EAP 6). To do this, run 
   $JBOSS_HOME/bin/standalone.sh

And so on. What do you think? I'd be happy just to see the note to check the directory after you run generate.fsh to see all the generated files. :-)

**Comment by sgilda** [comment]
Modified the README and issued pull request https://github.com/jbossas/quickstart/pull/84.
Once that one is verified and pushed, I think we can close this one.

**Comment by pmuir** [comment]
84 is closed, so I'm closing this one.

**Pull request closed without merging**